feat(clusters): Map cluster to ClickHouse cluster

lynnagara opened the pull request:
Clusters in Snuba can now be mapped to a physical ClickHouse cluster,
and they now aware of the set of server nodes that make up the cluster.

The cluster definition provided via settings (currently hardcoded to one
global cluster but soon to be configurable and allow more than one) will
take three additional parameters:
- single_node: Distributed vs local mode will no longer be a global
setting but a configurable property of a cluster. This allows us to use
a mix of local and distributed tables in production for different use
cases.
- cluster_name: This applies only to clusters with single_node = False,
and needs to match the ClickHouse cluster name defined in
/etc/clickhouse-server/config.xml.
- distributed_cluster_name: Also only applies to clusters with single_node
= False. If provided, the distributed hosts need to be registered as a
separate cluster in /etc/clickhouse-server/config.xml matching this name.

The cluster exposes the get_local_nodes() and get_distributed_nodes() methods
that provides a list of all of the server nodes containing the local and distributed
tables. In local mode, there is only one local node and no distributed nodes.
In distributed mode, the list of servers is found via the system.clusters table
where they must be registered.

These new methods are not currently used anywhere (and should not be
referenced anywhere until we have updated settings.CLUSTERS in all relevant
environments). In the future, this will be used in order to run certain
scripts (such as migrations, optimize, cleanup) that must be performed
on every shard not just the distributed table.

lynnagara committed:
feat(clusters): Map cluster to ClickHouse cluster

Clusters in Snuba can now be mapped to a physical ClickHouse cluster,
and they now aware of the set of server nodes that make up the cluster.

The cluster definition provided via settings (currently hardcoded to one
global cluster but soon to be configurable and allow more than one) will
take two additional parameters:
- distributed: Distributed vs local mode will no longer be a global
setting but a configurable property of a cluster. This allows us to use
a mix of local and distributed tables in production for different use
cases.
- cluster_name: This applies only to clusters with distributed = True,
and needs to match the ClickHouse cluster name defined in
/etc/clickhouse-server/config.xml.

The cluster exposes a get_nodes() method that returns all of the server
nodes of the cluster. In local mode, there is only one. In distributed
mode, the list of servers are found via the system.clusters table.

The get_nodes() method is not currently used anywhere (and should not be
referenced anywhere until we have updated settings.CLUSTERS in all relevant
environments). In the future, this will be used in order to run certain
scripts (such as migrations, optimize, cleanup) that must be performed
on every shard not just the distributed table.

lynnagara committed:
update test

lynnagara committed:
Merge remote-tracking branch 'origin/master' into cluster-hosts

lynnagara committed:
i give up

lynnagara commented on tests/clusters/test_cluster.py:
```diff
@@ -0,0 +1,64 @@
+        importlib.reload(cluster)
+
+    def teardown_class(self) -> None:
+        importlib.reload(settings)
+        importlib.reload(cluster)
```
This is really annoying, but I don't have an alternative idea for how to restore settings / cluster modules here.

fpacifici commented on snuba/clusters/cluster.py:
```diff
@@ -10,6 +13,14 @@
 
 
+@dataclass(frozen=True)
+class ClickhouseHost:
```
Aren't we supposed to distinguish between storage nodes and query nodes?
- is such parameter supposed to show up here?
- how would it be populated? Are we getting it from clickhosue or from the config, and if it comes from the config, how does that match with the configuration loaded from clickhouse ?

lynnagara replied:
Renamed to make it clearer that we are referring to the storage nodes here. This comes from either `system.clusters` in a multi node cluster, and is the same as the provided host and port (the query node) in a single node cluster.

fpacifici replied:
Ok, but I am not sure I get where we would configure the query nodes.  Are they going to be part of the cluster? The reader should come from one of the query nodes.

fpacifici commented on snuba/clusters/cluster.py:
```diff
@@ -61,16 +72,29 @@ class ClickhouseCluster(Cluster[SqlQuery, ClickhouseWriterOptions]):
+        port: int,
+        http_port: int,
+        storage_sets: Set[str],
+        is_distributed: bool,
```
I am not sure the `is_distributed` name is appropriate. We may have replication only, distributed may imply that we distribute data across nodes which means sharding.
`single_node` may be a better name to convey this meaning.

Taking a step back, I am not sure I am fully confident about which specific cases we need concept of distributed mode for. I remember only this:
- table name (which would be different)

Am I correct ?

lynnagara replied:
Right now it's only table name. In future, the distributed (or non single_node) option will also be used to determine the set of storage nodes that would be used for any of the commands that will need to also operate on them (cleanup/optimize/migrate).

fpacifici replied:
"will also be used to determine the set of storage nodes that would be used for any of the commands that will need to also operate on them"

I don't think I have clear how this will work. If you have a set of storage nodes and a set of query nodes in your cluster why do you need this flag to determine the set of storage nodes you would run those commands?

lynnagara replied:
What do you mean by "set of query nodes in your cluster"? As far as Snuba application is concerned I think there is only one query node per cluster (this is going to be the node with the distributed table in the distributed mode). Right now we only have one global query node, but we are moving towards one per cluster.

The idea here is that the user would only specify the host and port of the query node for the cluster. The storage nodes are discovered by going to the `system.clusters` table on the query node if it's a distributed cluster. Otherwise the storage node is just the query node if it's not distributed.

fpacifici replied:
I think there can be multiple query nodes in a distributed environment for a given clusters, on different hosts. Not all Clickhouse queries need to go through the same physical host. These nodes strictly speaking are not part of the clickhosue cluster (they do not show up in clickhouse configuration) in clickhouse terminology. 
Were you planning to expose the fact that there may be physically be multiple query nodes or to hide this in snuba ?

lynnagara replied:
After giving this some more thought, I think I've changed my mind about the need to distinguish storage and query nodes - I don't think a Snuba cluster needs to know about more than one node which is used for both storage and querying at all.

For a SELECT query, Snuba only ever connects to the one node in Clickhouse (reader always comes this). It's Clickhouse's job to either just run that (if it's local table) or figure out which of the one or combination of remote servers should actually run the query based on  sharding key, settings, Clickhouse configuration, etc (if it's distributed table).

INSERTs are the same. Again Snuba only knows about the one node with the distributed table. It's Clickhouse's job to figure out which actual servers need to get updated with the inserted data.

So the `get_nodes()` method introduced here is neither for storage nor query. It's for performing specific operations (like migrations) where we must actually connect to all of the individual hosts, so we need to discover what they are.

Does this make sense to you? I know it's different from what we discussed earlier about storage nodes vs query nodes, but I can't really remember the reason why that distinction was needed anymore.

fpacifici replied:
It seems to me your proposal is not really different from the original design. If a cluster contains a set of nodes used for DDL operations like migrations and still define one node used for insert and query, isn't this last node the query node? It still needs to be identified separately in the cluster configuration or in the snuba configuration so a distinction still needs to happen.

There is a secondary problem though (in both this proposal and the original design). How do we deal with multiple query nodes? We can do that to distribute the load. Are you planning to support this at all ? Or is snuba going to just know about one node to send query to ?
If we expose the existence of multiple nodes snuba will need to take care of load balancing (which it does not today). If we expose only one we need to give directions on how to do load balancing.

lynnagara replied:
Yeah, I'm suggesting something close to what we have now.

Are you proposing moving load balancing responsibilities into Snuba as part of this? That is not even something I'd considered at all (or really want to unless there's a compelling reason). What do you think the benefits of that would be would be? I was under the impression Clickhouse generally does a good job of this.

I think the only directions we give today are "in_order" or "random", is there some other directions you think would be required here even if we stick with the exposing single node approach?

fpacifici replied:
oh no, I don't think we should conflate clickhouse load balancing here, but we need to decide which directions to give. We could make snuba only be aware of one query node, this node can be a real host or a proxy that redirects to multiple clickhosue distributed table nodes so snuba does not need to know the existence of multiple clickhouse nodes with the distributed table deployed.
Though we need to be deliberate here.

The load balancer I am talking about is the one between snuba and clickhosue query nodes, not the one between clickhosue query nodes and clickhouse storage nodes.

lynnagara replied:
I added some documentation about this to the class. Let me know if it can also be clearer through the design.

fpacifici replied:
This approach should also be shared with the team in the design document, or with an amended document (probably there is no need for a meeting).

fpacifici replied:
have you verified how we would run the distributed table DDL operations? If snuba does not see all clickhouse nodes and they are all hidden behind one address snuba cannot be certain it created the distributed table on all nodes.

lynnagara replied:
I think there's 2 options to find those nodes.
1. Tell each Snuba cluster about those hosts via settings
2. Add that mapping to the clickhouse config and tell each Snuba cluster about that Clickhouse "cluster_name" so it can use that to look up those hosts

fpacifici replied:
`Add that mapping to the clickhouse config ` I don't think those nodes are in the clickhouse config. I think the config option is better now the questions are:
1) are these nodes part of the cluster data structure? If not how are they represented in terms of data structures? How does a migration script deal with them when running a DDL operation?
2) does the snuba cluster need to know them or is it just for the script running DDL operations.

lynnagara replied:
Yes, they'll need to be part of the cluster (and accessed by the script running the DDL operation). Even though it's not in the config yet, i'll add some code here showing how I think it can work.

fpacifici commented on snuba/clusters/cluster.py:
```diff
@@ -95,13 +119,27 @@ def get_writer(
+                )
+            ]
+        else:
+            return [ClickhouseHost(self.__host, self.__port)]
```
Why do we need to instantiate the cluster passing the `distributed` flag? Can't we infer whether we are distributed mode or not from the clickhouse response to the query you run on system.clusters ?

lynnagara replied:
The `system.clusters` table is not reliable when distributed=False, since there is a bunch of boilerplate in there. The distributed flag in this case isn't actually particularly useful and could be dropped. We can infer it from whether a user passes in a cluster_name, but I added it to be a bit more explicit since it might be easy for the user to inadvertently omit the cluster name. The "cluster_name" is the one that actually provides the name that is looked up in the system.clusters table.

fpacifici replied:
ok

lynnagara committed:
get_nodes -> get_storage_nodes

lynnagara committed:
Merge remote-tracking branch 'origin/master' into cluster-hosts

lynnagara committed:
rename distributed to single_node

lynnagara committed:
update test

lynnagara committed:
update test

lynnagara committed:
Merge remote-tracking branch 'origin/master' into cluster-hosts

lynnagara committed:
fix bad merge

lynnagara committed:
Merge branch 'master' into cluster-hosts

lynnagara committed:
try renaming again

lynnagara committed:
feat: Configure clusters via settings

Clusters are now configured via settings. The CLICKHOUSE_SERVER,
CLICKHOUSE_HOST, CLICKHOUSE_PORT and CLICKHOUSE_HTTP_PORT settings
are now deprecated in favor of cluster specific values provided
via the CLUSTERS list.

lynnagara committed:
Merge remote-tracking branch 'origin/master' into clusters-settings

lynnagara committed:
Merge remote-tracking branch 'origin/master' into cluster-hosts

lynnagara committed:
add comment about clickhouse nodes

lynnagara committed:
Merge remote-tracking branch 'origin/master' into cluster-hosts

fpacifici commented on snuba/clusters/cluster.py:
```diff
@@ -96,14 +107,39 @@ def get_writer(
-    shared by all storages located on the cluster
+    shared by all storages located on the cluster.
+
+    ClickhouseCluster is initialized with a single address (host/port/http_port),
+    which is used for all read and write operations related to the cluster. This
+    address can refer to either the address of the actual ClickHouse server, or a
+    proxy server (e.g. for load balancing).
```
Maybe add a couple of lines to explain the topology of the clickhosue cluster with regards to snuba: snuba nodes that run queries, clickhouse hosts that have dist tables that are hidden behind one address and clickhosue nodes that store data.

fpacifici reviewed (commented):
an additional comment in line

fpacifici commented on snuba/clusters/cluster.py:
```diff
@@ -148,13 +184,27 @@ def get_writer(
         )
 
+    def get_nodes(self) -> Sequence[ClickhouseNode]:
+        if self.__single_node:
```
Why not just storing the list of ClickhouseNode in the constructor and return that list? If we are in single_node mode we return one node only. It seems to me there is no need right now to store the __single_node field, so no need to add this to the state. 
If/when we will need it we will store it.

lynnagara replied:
Sure, I mean this function isn't being called anywhere right now so in a way it's not needed yet. Just putting in place some pieces that will be useful later. But I think it's more useful to store the single node since other things like table name would depend on that. It also avoids always making the unnecessary request to system.clusters in the constructor.

fpacifici commented on snuba/clusters/cluster.py:
```diff
@@ -96,14 +107,39 @@ def get_writer(
+        port: int,
+        http_port: int,
+        storage_sets: Set[str],
+        single_node: bool,
+        # The cluster name if single_node is set to False
+        cluster_name: Optional[str] = None,
```
We started the other long thread about the single_node issue and then it turned into a different topic (query vs storage nodes). Let's pick up the conversation on single_node here.

I don't think we need to pass the single node flag and store it for now (unless you plan to use it right away for the table name) it is not needed to return the proper list of nodes.

lynnagara replied:
I'm not sure I follow why it's not needed. Is is used to return the proper list of nodes - that either needs to come from the provided host parameter if single_node = true or `system.clusters` if single_node = false. It would also be used for the table name and anything else that currently depends on the global `DATASET_MODE` setting which would become deprecated in favor of cluster specific values.

fpacifici replied:
If you prepopulate the list of hosts at initialization you do not need to store it.
About the table name, my only concern is that I still have not clear who is supposed to do the mapping from an abstract name to the actual name on clickhouse. Is this going to be the responsibility of the Cluster object ?

lynnagara replied:
Right now table name is resolved by the table schema. As far as I'm aware it cannot be done by the storage since one storage may be backed by more than one physical table. I am not sure whether either 1) the schema should become aware of whether it's in the context of single node cluster or multi node cluster, or 2) this should be done somewhere else entirely.

lynnagara committed:
Add the distributed nodes

lynnagara committed:
fix

lynnagara commented:
I think we should do https://github.com/getsentry/snuba/pull/953 first, to avoid applying all the changes in production at once.

lynnagara committed:
Merge branch 'master' into clusters-settings

lynnagara committed:
Add is_single_node method

lynnagara commented:
@fpacifici any further thoughts on this?

fpacifici commented on snuba/clusters/cluster.py:
```diff
@@ -96,14 +107,42 @@ def get_writer(
         self.__host = host
         self.__port = port
         self.__http_port = http_port
+        self.__single_node = single_node
+        self.__cluster_name = cluster_name
+        self.__distributed_cluster_name = distributed_cluster_name
```
I am not sure I get why we need two cluster names since we have the single_node field. I was proposing alternatives to remove the single_node flag. But if we have it, can't we just have one cluster name ?

lynnagara replied:
Sorry if this part wasn't clear earlier. The `cluster_name` only gives the list of local servers. We'll run the local ddl on those machines. The second `distributed_cluster name` gives the list of hosts with the distributed tables. (this second part is new and needs to be added in Sentry prod) We'll run the distributed ddl on those machines. Does this approach make sense to you?

fpacifici reviewed (approved):
There is still a comment that is worth addressing, but not blocking, thanks.

lynnagara committed:
Merge remote-tracking branch 'origin/master' into cluster-hosts

lynnagara committed:
feat(clusters): Add the ability to connect with any node (#968)

Extend the get_connection method to add the ability to get a connection
with any node, not just the query node. Migrations will use this
connection in order to run specific commands on a node.

lynnagara committed:
Merge remote-tracking branch 'origin/clusters-settings' into cluster-hosts

lynnagara committed:
remove assertion

lynnagara committed:
Merge remote-tracking branch 'origin/master' into cluster-hosts

lynnagara committed:
Merge remote-tracking branch 'origin/master' into cluster-hosts
